### PR TITLE
Large online app refactor

### PR DESCRIPTION
**Pull request by Arden488**
A lot of stuff was rewritten.
Added a lot of animations.
Added a deck txt file and images.

### CODE REVIEW

**Review by hughwin — APPROVED** [review]

**Comment by hughwin** [comment]
I'll just give some instructions on how to get it working on your local machine as it didn't run fo me initially. 

- Start with the `mvn clean` `mvn package` goodness
- In the root folder, there us a TopTrumps.json file. Copy this to the target directory (where the .jar files are). 
- Start the jar with the command `java -jar TopTrumps.jar -o`
- Open your internet browser and connect to http://localhost:7777/toptrumps

**Comment by BelfDev** [comment]
> I'll just give some instructions on how to get it working on your local machine as it didn't run fo me initially.
> 
> * Start with the `mvn clean` `mvn package` goodness
> * In the root folder, there us a TopTrumps.json file. Copy this to the target directory (where the .jar files are).
> * Start the jar with the command `java -jar TopTrumps.jar -o`
> * Open your internet browser and connect to http://localhost:7777/toptrumps

We have a README @____@

**Comment by hughwin** [comment]
> 
> 
> > I'll just give some instructions on how to get it working on your local machine as it didn't run fo me initially.
> > 
> > * Start with the `mvn clean` `mvn package` goodness
> > * In the root folder, there us a TopTrumps.json file. Copy this to the target directory (where the .jar files are).
> > * Start the jar with the command `java -jar TopTrumps.jar -o`
> > * Open your internet browser and connect to http://localhost:7777/toptrumps
> 
> We have a README @____@

I know, but it doesn't mention moving the json file or connecting to the local host.

**Review by rjs68 — APPROVED** [review]
I don't understand all the code but there's nothing that jumps out as being obviously wrong - great stuff

**Comment by BelfDev** [comment]
Review in progress! Please don't merge.

**Comment by Arden488** [comment]
Fixed paths for libraries on the SelectionScreen and StatisticsScreen

File: src/main/resources/com/toptrumps/online/scripts/helpers/dom.js [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    const cardAttributeSelector = ".js-card-char";
+    const cardAttributeValueSelector = ".js-card-char-value";
+    const cardAttributesWrapperSelector = ".js-card-chars";
+    const endTurnButtonSelector = ".js-end-turn-button";
+    const nextTurnButtonSelector = ".js-modal-next-round-button";
+    const countdownSelector = ".js-countdown";
+    const cardImageSelector = ".js-card-image";
+    const cardTitleSelector = ".js-card-title";
+    const commonPileValueSelector = ".js-common-pile-value";
+    const commonPileSelector = ".js-common-pile";
```
Comment (BelfDev) [inline src/main/resources/com/toptrumps/online/scripts/helpers/dom.js:21]: communalPile :p

File: src/main/resources/com/toptrumps/online/scripts/models/player.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
+const PlayerModel = (($) => {
+    let players = {};
+    let playersList = {};
+
+    const init = playersData => {
+        playersData.forEach(player => {
```
Comment (BelfDev) [inline src/main/resources/com/toptrumps/online/scripts/models/player.js:6]: Theoretically, you could initiate players directly from the payload I get back to you :)

File: src/main/resources/com/toptrumps/online/scripts/models/player.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    };
+
+    const getTopCard = playerID => {
+        return players[playerID].deck.shift();
+    };
+
+    const getAiPlayersTopCards = () => {
+        const playersCards = {};
+
+        $.each(players, (i, player) => {
```
Comment (BelfDev) [inline src/main/resources/com/toptrumps/online/scripts/models/player.js:27]: Maybe ```$.filter()``` would be more adequate?

File: src/main/resources/com/toptrumps/online/scripts/models/player.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                name: players[data.playerID].name,
+                deckCount: players[data.playerID].deck.length + 1,
+                topCard: data.card
+            });
+        });
+
+        return dataset;
+    };
+
+    const getPlayersCardCount = () => {
```
Comment (BelfDev) [inline src/main/resources/com/toptrumps/online/scripts/models/player.js:52]: For every getter operation, you're having to traverse through a ```players``` array. Why not define a ```players``` variable inside your controller and extract all info you need at the beginning of each round -- so you can loop through it once.

File: src/main/resources/com/toptrumps/online/scripts/plugins/input-number.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            const data = {
+                min: elements.$input.attr("min"),
+                max: elements.$input.attr("max"),
+            };
+
+            bindEvents(elements, data);
+        });
+    };
+
+    const bindEvents = (elements, data) => {
```
Comment (BelfDev) [inline src/main/resources/com/toptrumps/online/scripts/plugins/input-number.js:28]: Very nice design!

File: src/main/resources/com/toptrumps/online/scripts/plugins/message.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+const Message = (($) => {
+    const gameStatusSelector = '.js-game-status-content';
+    const gameStatusLogSelector = ".js-game-log";
+    const gameLetterSelector = ".game-status__letter";
+    const gameLetterClass = "game-status__letter";
+
+    const animateLog = () => {
```
Comment (BelfDev) [inline src/main/resources/com/toptrumps/online/scripts/plugins/message.js:7]: I think the use of the name "log" here can cause some confusion (if you take into account the CLI's logger). I'd suggest ```infoPanel```.

File: src/main/resources/com/toptrumps/online/scripts/plugins/modal.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+const Modal = (($) => {
+    /** VARIABLES AND CONSTANTS */
+    // Selectors
+    const modalSelector = ".js-modal";
+    const backdropSelector = ".js-backdrop";
+    const modalTitleSelector = ".js-modal-title";
+    const modalHintSelector = ".js-modal-hint";
+
+    // Templates
+    const backdropTemplate = '<div class="backdrop js-backdrop"></div>';
```
Comment (BelfDev) [inline src/main/resources/com/toptrumps/online/scripts/plugins/modal.js:10]: ```'<div class="backdrop js-backdrop"/>'```

@___@

**Review by BelfDev — APPROVED** [review]
I'm approving but please take a look at my comments :)

**Pull request merged**